docs/api/synchronization_library/barrier.md

```diff
@@ -23,26 +23,35 @@ cuda::std::barrier<> bb;
 cuda::barrier<cuda::thread_scope_block> c;
 ```
 
-The class template `barrier` may also be declared without initialization; a
+The class template `barrier` may also be declared without initialization in the `cuda::` namespace; a
 friend function `init` may be used to initialize the object.
 
 ```c++
 // Shared memory does not allow initialization.
 __shared__ cuda::barrier<cuda::thread_scope_block> b;
 
 init(&b, 1); // Use this friend function to initialize the object.
+/*
+namespace cuda {
+  template<thread_scope Sco, class CompletionF>
+  __host__ __device__ void init(barrier<Sco,CompletionF>* bar, std::ptrdiff_t expected);
+  template<thread_scope Sco, class CompletionF>
+  __host__ __device__ void init(barrier<Sco,CompletionF>* bar, std::ptrdiff_t expected, CompletionF completion);
```

Comment:
Why do we add two overloads of `init` instead of one with a default `CompletionF` ? 

```suggestion
  template<thread_scope Sco, class CompletionF>
  __host__ __device__ void init(barrier<Sco,CompletionF>* bar, std::ptrdiff_t expected, CompletionF completion = CompletionF());
```

Reply:
I agree with Gonzalo that we should *document* this with just a single overload with `completion` defaulted.

The reason for why this is *implemented* as two overloads is that the friend functions are instantiated when you instantiate the class template, so that they can be found by ADL; and if `CompletionF` does not support default construction, the default argument is ill-formed and the entire program fails (even when you don't actually call `init`, if I remember how this works and what I encountered here correctly).

Reply:
That makes sense. 

The standard requires the default type of `CompletionF` to be `DefaultConstructible`, http://eel.is/c++draft/thread.barrier#class-6 , but it allows users to specify a non-`DefaultConstructible` one.